app/controllers/ValidationTaskController.scala

```diff
@@ -0,0 +1,160 @@
+
+  case class ValidationTaskPostReturnValue(mission: Option[Mission], labels: Option[JsValue])
+
+  /**
+    * Parse submitted validation data and submit to tables
+    * Useful info: https://www.playframework.com/documentation/2.6.x/ScalaJsonHttp
+    * BodyParsers.parse.json in async
+    */
+  def post = UserAwareAction.async(BodyParsers.parse.json) { implicit request =>
+    var submission = request.body.validate[Seq[ValidationTaskSubmission]]
+    println("[ValidationTaskController] submission: " + submission)
+    submission.fold(
+      errors => {
+        println("[ValidationTaskController] Error: " + errors);
+        Future.successful(BadRequest(Json.obj("status" -> "Error", "message" -> JsError.toFlatJson(errors))))
+      },
+      submission => {
+        val user = request.identity
+        val returnValues: Seq[ValidationTaskPostReturnValue] = for (data <- submission) yield {
+          for (interaction: InteractionSubmission <- data.interactions) {
+            ValidationTaskInteractionTable.save(ValidationTaskInteraction(0, interaction.missionId, interaction.action,
+              interaction.gsvPanoramaId, interaction.lat, interaction.lng, interaction.heading, interaction.pitch,
+              interaction.zoom, interaction.note, new Timestamp(interaction.timestamp)))
+          }
+
+          for (label: LabelValidationSubmission <- data.labels) {
+            user match {
+              case Some(user) =>
+                LabelValidationTable.save(LabelValidation(0, label.labelId, label.validationResult,
+                  user.userId.toString, label.missionId, label.canvasX, label.canvasY, label.heading,
+                  label.pitch, label.zoom, label.canvasHeight, label.canvasWidth,
+                  new Timestamp(label.startTimestamp), new Timestamp(label.endTimestamp)))
+              case None =>
+                Logger.warn("User without user_id validated a label, but every user should have a user_id.")
+            }
+          }
+
+          val missionId: Int = data.missionProgress.missionId
+          println("missionId: " + missionId)
+          val possibleNewMission: Option[Mission] = updateMissionTable(user, data.missionProgress)
+
+          // Temporary value - this is the number of labels that are in each mission.
+          val labelList: Option[JsValue] = getLabelList(data.missionProgress)
+          println("[ValidationTaskController] possibleMission: " + possibleNewMission)
+          println("[ValidationTaskController] labelList      : " + Json.toJson(labelList))
+          ValidationTaskPostReturnValue(possibleNewMission, labelList)
+        }
+        println("[ValidationTaskController] Success");
+        Future.successful(Ok(Json.obj(
+          "mission" -> returnValues.head.mission.map(_.toJSON),
+          "labels" -> returnValues.head.labels
+        )))
+      }
+    )
+  }
+
+  /**
+    * This function gets the metadata for a specific label in the database.
```

Comment:
The most nitpicky of points, but a lot of your comments start with "This function..." or "This method..." but that doesn't add anything. You don't actually have to change that everywhere, I just feel compelled to say it for some reason b/c I've read it so many times :joy:

Reply:
Heh whooops (you're totally right)! I think I changed some of my comments as I've been going through other code.